clients/gtest/blas1/asum_gtest.cpp

```diff
@@ -0,0 +1,158 @@
+ * PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
+ * CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * ************************************************************************ */
+#include "blas1_gtest.hpp"
+
+#include "hipblas_data.hpp"
+#include "hipblas_test.hpp"
+#include "testing_asum.hpp"
+#include "testing_asum_batched.hpp"
+#include "testing_asum_strided_batched.hpp"
+#include "type_dispatch.hpp"
+
+namespace
+{
+    // ----------------------------------------------------------------------------
+    // BLAS1 testing template
+    // ----------------------------------------------------------------------------
+    template <template <typename...> class FILTER, blas1 BLAS1>
+    struct asum_test_template : public HipBLAS_Test<asum_test_template<FILTER, BLAS1>, FILTER>
+    {
+        template <typename... T>
+        struct type_filter_functor
+        {
+            bool operator()(const Arguments& args)
+            {
+                // additional global filters applied first
+                if(!hipblas_client_global_filters(args))
+                    return false;
+
+                // type filters
+                return static_cast<bool>(FILTER<T...>{});
+            }
+        };
+
+        // Filter for which types apply to this suite
+        static bool type_filter(const Arguments& arg)
+        {
+            return hipblas_blas1_dispatch<asum_test_template::template type_filter_functor>(arg);
+        }
+
+        // Filter for which functions apply to this suite
+        static bool function_filter(const Arguments& arg);
+
+        // Google Test name suffix based on parameters
+        static std::string name_suffix(const Arguments& arg)
+        {
+            HipBLAS_TestName<asum_test_template> name(arg.name);
+            name << hipblas_datatype2string(arg.a_type);
+
+            if(strstr(arg.function, "_bad_arg") != nullptr)
+            {
+                name << "_bad_arg";
+            }
+            else
+            {
+                constexpr bool is_batched = (BLAS1 == blas1::asum_batched);
+                constexpr bool is_strided = (BLAS1 == blas1::asum_strided_batched);
+
+                name << '_' << arg.N << '_' << arg.incx;
+
+                if(is_strided)
+                {
+                    name << '_' << arg.stride_x;
+                }
+
+                if(is_batched || is_strided)
+                {
+                    name << "_" << arg.batch_count;
+                }
+            }
+
+            if(arg.api == FORTRAN)
+            {
+                name << "_F";
+            }
+
+            return std::move(name);
```

Comment:
Always thought if we were starting from scratch I would do this instead with a class much like ArgumentModel logging although maybe not variadic constructor.  As you are starting from scratch, might be worth a look although following rocblas also has it's benefits.

Reply:
I think this is a good idea, I'll draft something up in another PR after this goes in.

Reply:
> I think this is a good idea, I'll draft something up in another PR after this goes in.

I think I already wrote one for hipBLAS I had just forgot  :-) ,  rocBLAS we don't have one yet
class ArgumentModel
    void test_name(const Arguments& arg, std::string& name)
    
I was using that for the yaml tests I added for smoke testing I presume.... maybe you just need to hook into that form.

Reply:
Ha, even better; I'll make some further changes here then.